src/statistics/data.jl

```diff
@@ -48,21 +48,91 @@ quantile(d, p::T, s::Number) where {T<:Union{Number,AbstractVector}} = quantile(
 Spatial quantile of spatial data `sdata` at probability `p`.
 Optionally, specify the variable `v` and the block side `s`.
 """
-quantile(d::Data, p) = quantile(d, p, median_heuristic(d))
+quantile(d::Data, p) = quantile(d, p, mode_heuristic(d))
 
-function median_heuristic(d)
+function dist_matrix_random_sample(d, npoints=1000)
   # select at most 1000 points at random
   nel = nelements(d)
-  inds = sample(1:nel, min(nel, 1000), replace=false)
+  inds = sample(1:nel, min(nel, npoints), replace=false)
   X = (coordinates(centroid(d, ind)) for ind in inds)
-  D = pairwise(Euclidean(), X)
+  pairwise(Euclidean(), X)
+end
+
+bound_box_constr(d) = minimum(sides(boundingbox(domain(d))))
 
+function median_heuristic(d)
+  D = dist_matrix_random_sample(d)
   # median heuristic
   n = size(D, 1)
   m = median(D[i,j] for i in 1:n for j in 1:n if i > j)
 
-  # bounding box constraint
-  l = minimum(sides(boundingbox(domain(d))))
+  l = bound_box_constr(d)
 
   min(m, l)
 end
+
+"""
+    mode_heuristic(d)
+
+Return the estimated mode of the inter-point distances for a set of locations.
+"""
+function mode_heuristic(d)
+  D = dist_matrix_random_sample(d)
+  n = size(D, 1)
+  δ = [D[i,j] for i in 1:n for j in 1:n if i > j]
+  m = mode_hsm(δ)
+  l = bound_box_constr(d)
+  min(m, l)
+end
+
+"""
+    mode_hsm(x)
+    
+Return the mode of the vector `x`.
+
+## References
+
+* Bickel & Frühwirth, 2005. [On a fast, robust estimator
+  of the mode: Comparisons to other robust estimators
+  with applications](https://doi.org/10.1016/j.csda.2005.07.011)
+"""
+function mode_hsm(x_n)
+  sort!(x_n)
+  while length(x_n) ≥ 4
+    # find interval that contains approx 1/2 of data that has the smallest range
+    # indices n and k are constructed such that length(inf) == length(sup)
+    n = length(x_n)
+    k = trunc(Int, ceil(n / 2) - 1)
+    inf = x_n[1:(n - k)]
+    sup = x_n[(k + 1):n]
+    diffs = sup - inf
+    i = argmin(diffs)
+    if diffs[i] == 0
+      # if difference is zero, many points have the same value and we have found the mode
+      x_n = [x_n[i]]
+    else
+      # otherwise, take set with minimum range over n / 2 interval and continue to next
+      # halving iteration
+      x_n = x_n[i:(i+k)]
+    end
+  end
```

Comment:
Is this a classical bisection algorithm? Should we just remove all comments and state at the top of the while loop that this is a bisection to find median in a sorted vector?

Reply:
Yes, I think it is a bisecting search to find the "tallest" bin. 
Though I don't see an obvious way of restating the problem as a root finding problem with a sign change to get a total correspondence with the classical bisection method. 

Maybe there is a connection through calculus since the mode is the steepest part of the CDF?

Reply:
Thanks for the updates! I will try to read the reference you cited on this mode search algorithm. Maybe there is a way to simplify it a bit, I don't know.